thread-manager/examples/core_contention_sweep.rs

```diff
@@ -0,0 +1,216 @@
+    let mut all_results: HashMap<String, Results> = HashMap::new();
+    for regime in Regime::VALUES {
+        let mut results = Results::default();
+        for core_count in [2, 4, 8, 16] {
+            let manager;
+            info!("===================");
+            info!("Running {core_count} cores under {regime:?}");
+            let (tokio1, tokio2) = match regime {
+                Regime::Shared => {
+                    manager = ThreadManager::new(make_config_shared(core_count)).unwrap();
+                    (manager.get_tokio("axum1"), manager.get_tokio("axum2"))
+                }
+                Regime::Dedicated => {
+                    manager = ThreadManager::new(make_config_dedicated(core_count)).unwrap();
+                    (manager.get_tokio("axum1"), manager.get_tokio("axum2"))
+                }
+                Regime::Single => {
+                    manager = ThreadManager::new(make_config_shared(core_count)).unwrap();
+                    (manager.get_tokio("axum1"), manager.get_tokio("axum2"))
+                }
+            };
+
+            let wrk_cores: Vec<_> = (32..64).collect();
+            let measurement = std::thread::scope(|s| {
+                s.spawn(|| {
+                    tokio1.start_metrics_sampling(Duration::from_secs(1));
+                    tokio1.tokio.block_on(axum_main(8888));
+                });
+                let jh = match regime {
+                    Regime::Single => s.spawn(|| {
+                        run_wrk(&[8888, 8888], &wrk_cores, wrk_cores.len(), 3000).unwrap()
+                    }),
+                    _ => {
+                        s.spawn(|| {
+                            tokio2.start_metrics_sampling(Duration::from_secs(1));
+                            tokio2.tokio.block_on(axum_main(8889));
+                        });
+                        s.spawn(|| {
+                            run_wrk(&[8888, 8889], &wrk_cores, wrk_cores.len(), 3000).unwrap()
+                        })
+                    }
+                };
+                jh.join().expect("WRK crashed!")
+            });
+            info!("Results are: {:?}", measurement);
+            results.latencies_s.push(
+                measurement.0.iter().map(|a| a.as_secs_f32()).sum::<f32>()
+                    / measurement.0.len() as f32,
+            );
+            results.rps.push(measurement.1.iter().sum());
+        }
+        all_results.insert(format!("{regime:?}"), results);
+        std::thread::sleep(Duration::from_secs(3));
+    }
+
+    //print the resulting measurements so they can be e.g. plotted with matplotlib
+    println!("{}", serde_json::to_string_pretty(&all_results)?);
+
+    Ok(())
+}
+
+fn run_wrk(
+    ports: &[u16],
+    cpus: &[usize],
+    threads: usize,
+    connections: usize,
+) -> anyhow::Result<(Vec<Duration>, Vec<f32>)> {
+    //Sleep a bit to let axum start
+    std::thread::sleep(Duration::from_millis(500));
```

Comment:
😭

Reply:
Added proper signals to avoid this nightmare.